_posts/en/2016-12-28-delegates-closures.md

```diff
@@ -0,0 +1,299 @@
+    manager.delegate = self;
+    [manager get:@"http://plop.fr/json"];
+}
+
+- (void)onRequestFailure {
+
+}
+
+- (void)onRequestSuccess {
+
+}
+
+@end
+```
+
+```Objective-C
+@interface RequestManager : NSObject
+{
+
+}
+
+@property (nonatomic, weak) id delegate;
+
+- (void)get:(NSString*)url;
+```
+
+```Objective-C
+#import "RequestManager.h"
+
+@implementation RequestManager
+
+- (void)get:(NSString *)url {
+    //Do the call
+    BOOL requestSucceed = [self isSuccess];
+
+    //After the call
+    if (requestSucceed) {
+        [self.delegate onRequestSuccess];
+    } else {
+        [self.delegate onRequestFailure];
+    }
+}
+
+- (BOOL)isSuccess {
+    return YES;
+}
+
+@end
+```
+
+##### Swift
+
+```Swift
+class MyClassSwift: UIViewController, RequesterDelegateSwift {
+
+    func callWebService() {
+        let manager: RequestManager = RequestManager()
+        manager.delegate = self
+        manager.get(url: "http://plop.fr/json")
+    }
+
+    func onRequestFailure() {
+
+    }
+
+    func onRequestSuccess() {
+    }
+}
+```
+
+```Swift
+class RequestManager {
+
+    var delegate: RequesterDelegateSwift?
+
+    func get(url: String) {
+        //Do the call
+        let requestSucceed: Bool = self.isSuccess()
+
+        //After the call
+        if requestSucceed {
+            self.delegate?.onRequestSuccess()
+        } else {
+            self.delegate?.onRequestFailure()
+        }
+    }
+
+    private func isSuccess() -> Bool {
+        return true
+    }
+}
+```
+
+**But why do all this?**
+
+We have a light dependency between our different objects. Indeed, RequestManager has no idea of the type of the object of its delegate, all that it cares about is to know that the object contains the two methods defined in the protocol in order to be able to call them. So, It's good, I was able to call my webservice, I have my callback methods in which I can handle the different cases, everything seems good to me. Yes, everything is good, it works well and it will cover a lot of cases
+
+**Raised Issue**
+
+- But now, if we need to call many webservices in only one class, how do we do?
+- How do we do? You just showed us how to do it.
+- The issue is that by doing it this way, all the returns of your webservices will go in the your methods onRequestSuccess and onRequestFailure.
+- Ha, but you are right, it raises an issue...
+
+**Solution**
+
+In order to solve this issue, there is still the possibility to change a bit the method, add a unique id when we launch the request and to receive it when the request is done and they identify to which request it belongs to. Yes it is true, we can do that, but it will fast become "verbose" and not so easy to handle. One thing I like to do when I program is to avoid strong dependency (Yes, the delegate already helps for that), but even more, I like when I can move some code really easily. But here, If I want to move my code, I need to re-implement the protocol in one other class.
```

Comment:
and receive it
quickly become
helps that
re-implement the protocol in another class